hash/hash.go

```diff
@@ -0,0 +1,127 @@
+package hash
+
+import (
+	"bytes"
+	"crypto/subtle"
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/NebulousLabs/fastrand"
+	"golang.org/x/crypto/argon2"
+)
+
+var (
+	ErrInvalidHash               = errors.New("the encoded hash is not in the correct format")
+	ErrIncompatibleVersion       = errors.New("incompatible version of argon2")
+	ErrMismatchedHashAndPassword = errors.New("passwords do not match")
+
+	// config is the configuration of the argon2id hasher.
+	// For the moment we'll use the setup we had with Kratos. Later on we can
+	// move this to a configuration file, if needed.
+	config = argon2Config{
+		SaltLength:  16,
+		Iterations:  2,
+		Memory:      131072,
+		Parallelism: 1,
+		KeyLength:   16,
```

Comment:
We could. These are the default values used by Kratos.

The main advantage of the argon2 algo is that the hash size is not the only determining factor in determining the complexity of the hashing process. We want slow hashing, so people won't be able to brute force it but not so slow that it would be bad for us as well. This is controlled by the other 3 parameters.

16 bytes is the recommended size by argon2's authors but I don't see a problem with doubling that.

Links:
 - https://www.ory.sh/choose-recommended-argon2-parameters-password-hashing/
 - https://password-hashing.net/argon2-specs.pdf (see "Recommended parameters")

If we choose to change the length, all existing passwords won't be affected.